File: packages/core/src/protocol.ts
export const PROTOCOL_VERSION = 1 as const;

export type RequestId = string;
export type PageId = number;
export type WindowId = number;

export interface LogicalRect {
  x: number;
  y: number;
  width: number;
  height: number;
}

export type CoordinateSpace = "frontend_logical";

export interface PageViewBounds {
  rect: LogicalRect;
  coordinate_space: CoordinateSpace;
  device_pixel_ratio: number;
  visual_viewport_scale: number;
}

export interface PageViewBoundsInput {
  rect: LogicalRect;
  coordinateSpace: CoordinateSpace;
  devicePixelRatio: number;
  visualViewportScale: number;
}

export interface PageOptions {
  url: string;
}

export interface PageState {
  id: PageId;
  url: string;
  title: string;
  loading: boolean;
  can_go_back: boolean;
  can_go_forward: boolean;
  active: boolean;
}

export interface WindowState {
  id: WindowId;
  active_page_id: PageId | null;
  pages: PageState[];
}

export type Command =
  | { type: "page_navigate"; page_id: PageId; url: string }
  | { type: "page_reload"; page_id: PageId }
  | { type: "page_go_back"; page_id: PageId }
  | { type: "page_go_forward"; page_id: PageId }
  | { type: "page_stop"; page_id: PageId }
  | { type: "page_activate"; window_id: WindowId; page_id: PageId }
  | { type: "page_close"; window_id: WindowId; page_id: PageId }
  | { type: "page_register_view"; page_id: PageId }
  | ({ type: "page_set_view_bounds"; page_id: PageId } & PageViewBounds)
  | { type: "page_unregister_view"; page_id: PageId };

export type Request =
  | { type: "runtime_handshake"; protocol_version: number }
  | { type: "page_create"; window_id: WindowId; options: PageOptions }
  | { type: "page_get_state"; page_id: PageId }
  | { type: "window_get_state"; window_id: WindowId };

export type FrontendMessage =
  | { type: "command"; command: Command }
  | { type: "request"; id: RequestId; request: Request };

export type ResponseData =
  | { type: "runtime_handshake"; protocol_version: number; windows: WindowState[] }
  | { type: "page_created"; page: PageState }
  | { type: "page_state"; page: PageState }
  | { type: "window_state"; window: WindowState };

export type ProtocolErrorCode =
  | "invalid_message"
  | "unsupported_message"
  | "protocol_version_mismatch"
  | "window_not_found"
  | "page_not_found"
  | "invalid_url"
  | "runtime_not_ready"
  | "runtime_shutting_down"
  | "request_failed"
  | "invalid_view_bounds";

export interface ProtocolError {
  code: ProtocolErrorCode;
  message: string;
}

export type NativeMessage =
  | { type: "response"; id: RequestId; result: { status: "ok"; data: ResponseData } | { status: "err"; error: ProtocolError } }
  | { type: "event"; event: Event };

export type Event =
  | { type: "page_created"; page: PageState }
  | { type: "page_url_changed"; page_id: PageId; url: string }
  | { type: "page_title_changed"; page_id: PageId; title: string }
  | { type: "page_loading_changed"; page_id: PageId; loading: boolean }
  | { type: "page_navigation_state_changed"; page_id: PageId; can_go_back: boolean; can_go_forward: boolean }
  | { type: "page_activated"; window_id: WindowId; page_id: PageId }
  | { type: "page_closed"; window_id: WindowId; page_id: PageId };

export function parseNativeMessage(value: unknown): NativeMessage {
  if (!value || typeof value !== "object") {
    throw new Error("native message must be an object");
  }
  const message = value as { type?: unknown; id?: unknown; result?: unknown; event?: unknown };
  if (message.type === "response" && typeof message.id === "string" && message.result && typeof message.result === "object") {
    const result = message.result as { status?: unknown; data?: unknown; error?: unknown };
    if (result.status === "ok" && result.data && typeof result.data === "object") return value as NativeMessage;
    if (result.status === "err" && result.error && typeof result.error === "object") return value as NativeMessage;
  }
  if (message.type === "event" && message.event && typeof message.event === "object") {
    return value as NativeMessage;
  }
  throw new Error("unknown native message type");
}

File: crates/browserkit-types/src/protocol.rs
use super::{
    BrowserEvent, CoordinateSpace, LogicalRect, PageId, PageOptions, PageState, WindowId,
    WindowState,
};
use serde::{Deserialize, Serialize};

pub const PROTOCOL_VERSION: u32 = 1;

#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
#[serde(transparent)]
pub struct RequestId(pub String);

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum FrontendMessage {
    Command(CommandEnvelope),
    Request(RequestEnvelope),
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct CommandEnvelope {
    pub command: Command,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct RequestEnvelope {
    pub id: RequestId,
    pub request: Request,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Command {
    PageNavigate {
        page_id: PageId,
        url: String,
    },
    PageReload {
        page_id: PageId,
    },
    PageGoBack {
        page_id: PageId,
    },
    PageGoForward {
        page_id: PageId,
    },
    PageStop {
        page_id: PageId,
    },
    PageActivate {
        window_id: WindowId,
        page_id: PageId,
    },
    PageClose {
        window_id: WindowId,
        page_id: PageId,
    },
    PageRegisterView {
        page_id: PageId,
    },
    PageSetViewBounds {
        page_id: PageId,
        rect: LogicalRect,
        coordinate_space: CoordinateSpace,
        device_pixel_ratio: f64,
        visual_viewport_scale: f64,
    },
    PageUnregisterView {
        page_id: PageId,
    },
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Request {
    RuntimeHandshake {
        protocol_version: u32,
    },
    PageCreate {
        window_id: WindowId,
        options: PageOptions,
    },
    PageGetState {
        page_id: PageId,
    },
    WindowGetState {
        window_id: WindowId,
    },
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum NativeMessage {
    Response(ResponseEnvelope),
    Event(EventEnvelope),
}

pub trait ProtocolSink: Send + Sync {
    fn send(&self, message: NativeMessage);
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ResponseEnvelope {
    pub id: RequestId,
    pub result: ResponseResult,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum ResponseResult {
    Ok { data: ResponseData },
    Err { error: ProtocolError },
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ResponseData {
    RuntimeHandshake {
        protocol_version: u32,
        windows: Vec<WindowState>,
    },
    PageCreated {
        page: PageState,
    },
    PageState {
        page: PageState,
    },
    WindowState {
        window: WindowState,
    },
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct EventEnvelope {
    pub event: Event,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Event {
    PageCreated {
        page: PageState,
    },
    PageUrlChanged {
        page_id: PageId,
        url: String,
    },
    PageTitleChanged {
        page_id: PageId,
        title: String,
    },
    PageLoadingChanged {
        page_id: PageId,
        loading: bool,
    },
    PageNavigationStateChanged {
        page_id: PageId,
        can_go_back: bool,
        can_go_forward: bool,
    },
    PageActivated {
        window_id: WindowId,
        page_id: PageId,
    },
    PageClosed {
        window_id: WindowId,
        page_id: PageId,
    },
}

impl From<BrowserEvent> for Event {
    fn from(event: BrowserEvent) -> Self {
        match event {
            BrowserEvent::PageUrlChanged { page_id, url } => Self::PageUrlChanged { page_id, url },
            BrowserEvent::PageTitleChanged { page_id, title } => {
                Self::PageTitleChanged { page_id, title }
            }
            BrowserEvent::PageLoadingChanged { page_id, loading } => {
                Self::PageLoadingChanged { page_id, loading }
            }
            BrowserEvent::PageNavigationStateChanged {
                page_id,
                can_go_back,
                can_go_forward,
            } => Self::PageNavigationStateChanged {
                page_id,
                can_go_back,
                can_go_forward,
            },
            BrowserEvent::PageActivated { window_id, page_id } => {
                Self::PageActivated { window_id, page_id }
            }
            BrowserEvent::PageClosed { window_id, page_id } => {
                Self::PageClosed { window_id, page_id }
            }
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ProtocolError {
    pub code: ProtocolErrorCode,
    pub message: String,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProtocolErrorCode {
    InvalidMessage,
    UnsupportedMessage,
    ProtocolVersionMismatch,
    WindowNotFound,
    PageNotFound,
    InvalidUrl,
    RuntimeNotReady,
    RuntimeShuttingDown,
    RequestFailed,
    InvalidViewBounds,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ids_use_stable_wire_values() {
        let message = FrontendMessage::Request(RequestEnvelope {
            id: RequestId("req-42".into()),
            request: Request::PageGetState {
                page_id: PageId::from_raw(7),
            },
        });
        let json = serde_json::to_string(&message).unwrap();
        assert!(json.contains("req-42"));
        assert!(json.contains("\"page_id\":7"));
        assert_eq!(
            serde_json::from_str::<FrontendMessage>(&json).unwrap(),
            message
        );
    }

    #[test]
    fn every_envelope_round_trips() {
        let message = NativeMessage::Event(EventEnvelope {
            event: Event::PageLoadingChanged {
                page_id: PageId::from_raw(1),
                loading: true,
            },
        });
        let json = serde_json::to_string(&message).unwrap();
        assert_eq!(
            serde_json::from_str::<NativeMessage>(&json).unwrap(),
            message
        );
    }

    #[test]
    fn protocol_version_is_single_source_of_truth() {
        assert_eq!(PROTOCOL_VERSION, 1);
    }

    #[test]
    fn view_bounds_keep_wire_names_and_fractional_values() {
        let message = FrontendMessage::Command(CommandEnvelope {
            command: Command::PageSetViewBounds {
                page_id: PageId::from_raw(1),
                rect: LogicalRect {
                    x: 12.5,
                    y: 24.0,
                    width: 640.25,
                    height: 480.0,
                },
                coordinate_space: CoordinateSpace::FrontendLogical,
                device_pixel_ratio: 2.0,
                visual_viewport_scale: 1.0,
            },
        });
        let json = serde_json::to_string(&message).unwrap();
        assert!(json.contains("page_set_view_bounds"));
        assert!(json.contains("frontend_logical"));
        assert_eq!(
            serde_json::from_str::<FrontendMessage>(&json).unwrap(),
            message
        );
    }
}
